tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java

```diff
@@ -200,6 +190,37 @@ static Properties readProps(List<String> producerProps, String producerConfig, S
         return props;
     }
 
+
+    static List<byte[]> generatePayloads(String payloadFilePath, String payloadDelimiter, Integer recordSize, Random random) throws IOException {
+        if (payloadFilePath != null) {
+            return readPayloadFile(payloadFilePath, payloadDelimiter);
+        } else if (recordSize != null) {
+            return generateRandomPayloads(recordSize, random);
+        } else {
+            throw new IllegalArgumentException("no payload File Path or record Size provided");
+        }
+    }
+
+
+    private static List<byte[]> generateRandomPayloads(final int recordSize, final Random random) {
+        if (recordSize == 0) {
+            final ArrayList<byte[]> bytes = new ArrayList<>(1);
+            bytes.add(new byte[0]);
+            return bytes;
+        }
+        int approximateTotalSize = 10 * 1024 * 1024;
+        int numberOfCandidates = Math.max(1, approximateTotalSize / Math.max(1, recordSize));
```

Comment:
Why not use `numRecords` from start() instead ?
Don't we run the risk of regressing `https://github.com/apache/kafka/pull/10469` if `numberOfCandidates` is less than `numRecords` ?

Reply:
If we pre-generate using numRecords we could run out of heap. Though I see your point, with this implementation the messages batches would become more compressible as record-size approaches 10M and there are less candidates generated.